#!/usr/bin/env Rscript

#==================================
# used for ATAC seq quality analysis
# adapted from the link:
# https://bioconductor.org/packages/release/bioc/vignettes/ATACseqQC/inst/doc/ATACseqQC.html

# check the code Install_Bioconductor_Packages_using_BioCLite.R
# to find out the required packages

#==================================
# author: Sourya Bhattacharyya
# Vijay-AY lab
#==================================

suppressPackageStartupMessages({
  library(ATACseqQC)
  library(ChIPpeakAnno)
  library(MotifDb)
  library(GenomicRanges)
  library(GenomicAlignments)
  library(optparse)  
})

#===========================================================
option_list = list(
	make_option(c("--AlignFile"), type="character", default=NULL, help="Input alignment file. Must contain an index file (generated by samtools)."),	
	make_option(c("--RefGenome"), type="character", default=NULL, help="Reference genome name."),
	make_option(c("--OutDirQC"), type="character", default=NULL, help="Output directory to contain the QC related statistics."),
	make_option(c("--PE"), type="integer", action="store", default=0, help="If 1, indicates paired end input data. Default = 0")
); 

opt_parser = OptionParser(option_list=option_list);
opt = parse_args(opt_parser);

# dynamic loading of libraries based on the reference genome
if (opt$RefGenome == 'hg19') {
	library(BSgenome.Hsapiens.UCSC.hg19)
	library(TxDb.Hsapiens.UCSC.hg19.knownGene)
	library(phastCons100way.UCSC.hg19)
} else if (opt$RefGenome == 'hg38') {
	library(BSgenome.Hsapiens.UCSC.hg38)
	library(TxDb.Hsapiens.UCSC.hg38.knownGene)
	library(phastCons100way.UCSC.hg38)
} else if (opt$RefGenome == 'mm10') {
	library(BSgenome.Mmusculus.UCSC.mm10)
	library(TxDb.Mmusculus.UCSC.mm10.knownGene)
	library(phastCons60way.UCSC.mm10)
} else if (opt$RefGenome == 'mm9') {
	library(BSgenome.Mmusculus.UCSC.mm9)
	library(TxDb.Mmusculus.UCSC.mm9.knownGene)
	library(phastCons60way.UCSC.mm9)
}


# create the output directory for QC
outdir <- opt$OutDirQC
system(paste("mkdir -p", outdir))

# input the bamFile from the ATACseqQC package 
# bamfile <- system.file("extdata", inpbamfilename, package="ATACseqQC", mustWork=TRUE)
bamfile <- opt$AlignFile
bamfile.labels <- gsub(".bam", "", basename(bamfile))

# store the current directory
currdir <- getwd()

# go to the directory "outdir"
setwd(outdir)

#=====================
# Check alignment metrics and mapping quality
bam_QC_textfile <- paste0(outdir, '/Summary_QC.txt')
if (file.exists(bam_QC_textfile) == FALSE) {
	fp_out <- file(bam_QC_textfile, "w")
	outtext <- paste0("\n *** Quality control measures for the alignment file ", opt$AlignFile, " is **** \n")
	writeLines(outtext, con=fp_out, sep="\n")
	close(fp_out)
	# quality control summary
	capture.output(bamQC(bamfile, outPath=NULL), file=bam_QC_textfile, append=TRUE)
	cat(sprintf("\n *** Computed the sumary statistics of the input ATAC seq file ****  \n"))
}

# #=====================
# Estimate the library complexity
# sourya - commented since old bioconductor version does not support this function
if (0) {
	plotfile <- paste0(outdir, '/Library_Complexity.pdf')
	pdf(plotfile, width=6, height=4)
	estimateLibComplexity(readsDupFreq(bamfile))
	dev.off()
	cat(sprintf("\n *** Computed library complexity ****  \n"))
}

#=====================
# works only when the input is single end
if (opt$PE == 0) {
	# shift the BAM file - forward strand by +4 bp
	possibleTag <- c("AS", "XN", "XM", "XO", "XG", "NM", "MD", "YS", "YT")
	gal <- readBamFile(bamfile, asMates=FALSE)
	shiftedBamfile <- paste0(outdir, '/shifted.bam')
	gal1 <- shiftGAlignmentsList(gal)
	export(gal1, shiftedBamfile)
	cat(sprintf("\n *** shifted bam file ****  \n"))
}

#=====================
# works only when the input is paired end
if (opt$PE == 1) {
	
	# fragment size distribution (main QC metric)
	plotfile <- paste0(outdir, '/Fragment_Size_Distribution.pdf')
	if (file.exists(plotfile) == FALSE) {
		pdf(plotfile, width=6, height=4)
		fragSizeDist(bamfile, bamfile.labels)
		dev.off()
		cat(sprintf("\n *** Computed the fragment size distribution (paired end read) **** \n"))
	}

	# prepare the tags
	# obtained from: 
	# https://bioinformatics-core-shared-training.github.io/cruk-summer-school-2019/ChIPSeq/Materials/Practicals/Day5/Practical01_ATAC-seq_analysis_SS.html
	# and from 
	# https://bioconductor.org/packages/release/bioc/vignettes/ATACseqQC/inst/doc/ATACseqQC.html

	# option 1: all combination
	# possibleTag <- combn(LETTERS, 2)
	# possibleTag <- c(paste0(possibleTag[1, ], possibleTag[2, ]), paste0(possibleTag[2, ], possibleTag[1, ]))
	# cat(sprintf("\n length of possibleTag : %s ", length(possibleTag)))

	# option 2: specified combinations
	possibleTag <- c("AS", "XN", "XM", "XO", "XG", "NM", "MD", "YS", "YT")

	# prepare the seqlev input
	# by default  seqlev = paste0("chr", c(1:22, "X", "Y"))
	# but it should be checked with the BAM header
	seqlevset_default <- paste0("chr", c(1:22, "X", "Y"))
	tempBAMHeaderFile <- paste0(outdir, '/t.bed')
	system(paste("samtools view -H ", bamfile, " | awk -F\'[\t]\' \'($1==\"@SQ\")\' - > ", tempBAMHeaderFile))
	nline <- as.integer(system(paste("cat", tempBAMHeaderFile, "| wc -l"), intern = TRUE))
	if (nline > 0) {
		x <- read.table(tempBAMHeaderFile, header=F, sep="\t", stringsAsFactors=F)
		y <- x[,2]
		z <- as.vector(unlist(strsplit(y, ":")))
		seqlevset_final <- intersect(seqlevset_default, z)
		cat(sprintf("\n ==>> chromosomes considered : %s ", paste(as.vector(seqlevset_final), sep="\t")))
	} else {
		seqlevset_final <- seqlevset_default
	}
	system(paste("rm", tempBAMHeaderFile))

	# compute promoter transcript body score
	if (opt$RefGenome == 'hg19') {
		txs <- transcripts(TxDb.Hsapiens.UCSC.hg19.knownGene)		
	} else if (opt$RefGenome == 'hg38') {
		txs <- transcripts(TxDb.Hsapiens.UCSC.hg38.knownGene)
	} else if (opt$RefGenome == 'mm10') {
		txs <- transcripts(TxDb.Mmusculus.UCSC.mm10.knownGene)
	} else if (opt$RefGenome == 'mm9') {
		txs <- transcripts(TxDb.Mmusculus.UCSC.mm9.knownGene)
	}
	cat(sprintf("\n ==>> read transcript information "))

	for (chrIdx in 1:length(seqlevset_final)) {
		currChr_seqlev <- seqlevset_final[chrIdx]
		cat(sprintf("\n extracting nucleosome specific BAM - chromosome : %s ", currChr_seqlev))		
		
		# extract current chromosome specific genome information
		if ((opt$RefGenome == 'hg19') | (opt$RefGenome == 'hg38')) {
			which_currChr <- as(seqinfo(Hsapiens)[currChr_seqlev], "GRanges")
		} else if ((opt$RefGenome == 'mm10') | (opt$RefGenome == 'mm9')) {
			which_currChr <- as(seqinfo(Mmusculus)[currChr_seqlev], "GRanges")
		}

		# read the BAM file for the current chromosome
		# and shift it in both ends
		gal <- readBamFile(bamfile, tag=possibleTag, which=which_currChr, asMates=TRUE)
		cat(sprintf("\n read bam file for the current chromosome : %s ", currChr_seqlev))

		# shift the GAlignmentsLists by 5' ends. 
		# All reads aligning to the positive strand will be offset by +4bp, 
		# and all reads aligning to the negative strand will be offset -5bp by default.
		shiftedBamfile <- file.path(outdir, 'shifted.bam')
		gal1 <- shiftGAlignmentsList(gal)
		export(gal1, shiftedBamfile)
		cat(sprintf("\n shifted bam file for the current chromosome : %s ", currChr_seqlev))

		# following code is to get the nucleosome specific
		# and nucleosome free regions
		# but this is giving some problems - comment - sourya
		if (0) {

		# extract transcripts for the current chromosome
		txs_currChr_seqlev <- txs[seqnames(txs) %in% currChr_seqlev]
		cat(sprintf("\n extracted transcripts for the current chromosome : %s ", currChr_seqlev))

		# call the split BAM routine for the current chromosome
		if (opt$RefGenome == 'hg19') {
			objs <- splitGAlignmentsByCut(gal1, txs=txs_currChr_seqlev, genome=Hsapiens, conservation=phastCons100way.UCSC.hg19)
		} else if (opt$RefGenome == 'hg38') {
			objs <- splitGAlignmentsByCut(gal1, txs=txs_currChr_seqlev, genome=Hsapiens, conservation=phastCons100way.UCSC.hg38)
		} else if (opt$RefGenome == 'mm10') {
			objs <- splitGAlignmentsByCut(gal1, txs=txs_currChr_seqlev, genome=Mmusculus, conservation=phastCons60way.UCSC.mm10)
		} else if (opt$RefGenome == 'mm9') {
			objs <- splitGAlignmentsByCut(gal1, txs=txs_currChr_seqlev, genome=Mmusculus, conservation=phastCons60way.UCSC.mm9)
		}
		cat(sprintf("\n created objs for the current chromosome : %s ", currChr_seqlev))

		# export the binned alignments into bam files.
		null <- writeListOfGAlignments(objs, outdir)
		cat(sprintf("\n extracted nucleosome specific reads for the current chromosome : %s ", currChr_seqlev))

		# now rename the files so that chromosome specific information is preserved
		system(paste("mv NucleosomeFree.bam", paste0("NucleosomeFree_", currChr_seqlev, ".bam")))
		system(paste("mv NucleosomeFree.bam.bai", paste0("NucleosomeFree_", currChr_seqlev, ".bam.bai")))
		system(paste("mv mononucleosome.bam", paste0("mononucleosome_", currChr_seqlev, ".bam")))
		system(paste("mv mononucleosome.bam.bai", paste0("mononucleosome_", currChr_seqlev, ".bam.bai")))
		system(paste("mv dinucleosome.bam", paste0("dinucleosome_", currChr_seqlev, ".bam")))
		system(paste("mv dinucleosome.bam.bai", paste0("dinucleosome_", currChr_seqlev, ".bam.bai")))
		system(paste("mv trinucleosome.bam", paste0("trinucleosome_", currChr_seqlev, ".bam")))
		system(paste("mv trinucleosome.bam.bai", paste0("trinucleosome_", currChr_seqlev, ".bam.bai")))
		system(paste("mv inter1.bam", paste0("inter1_", currChr_seqlev, ".bam")))
		system(paste("mv inter1.bam.bai", paste0("inter1_", currChr_seqlev, ".bam.bai")))
		system(paste("mv inter2.bam", paste0("inter2_", currChr_seqlev, ".bam")))
		system(paste("mv inter2.bam.bai", paste0("inter2_", currChr_seqlev, ".bam.bai")))
		system(paste("mv inter3.bam", paste0("inter3_", currChr_seqlev, ".bam")))
		system(paste("mv inter3.bam.bai", paste0("inter3_", currChr_seqlev, ".bam.bai")))
		system(paste("mv others.bam", paste0("others_", currChr_seqlev, ".bam")))
		system(paste("mv others.bam.bai", paste0("others_", currChr_seqlev, ".bam.bai")))

		}	# end dummy if
		# end code  comment - sourya

		# only rename the shifted files
		system(paste("mv shifted.bam", paste0("shifted_", currChr_seqlev, ".bam")))
		system(paste("mv shifted.bam.bai", paste0("shifted_", currChr_seqlev, ".bam.bai")))
	
	}	# end chromosome loop
	
	cat(sprintf("\n *** Split alignments ****  \n"))

	# sourya - commented since old bioconductor version does not support this function
	if (0) {
		pt <- PTscore(gal1, txs)
		plotfile <- paste0(outdir, '/promoter_transcript_body_score.pdf')
		pdf(plotfile, width=6, height=4)	
		plot(pt$log2meanCoverage, pt$PT_score, xlab="log2 mean coverage", ylab="Promoter vs Transcript")
		dev.off()
	}

	# Nucleosome Free Regions (NFR) score
	# sourya - commented since old bioconductor version does not support this function
	if (0) {	
		nfr <- NFRscore(gal1, txs)
		plotfile <- paste0(outdir, '/Nucleosome_Free_Regions_score.pdf')
		pdf(plotfile, width=6, height=4)
		plot(nfr$log2meanCoverage, nfr$NFR_score, xlab="log2 mean coverage", ylab="Nucleosome Free Regions score", main="NFRscore for 200bp flanking TSSs", xlim=c(-10, 0), ylim=c(-5, 5))
		dev.off()
	}

	# Transcription Start Site (TSS) Enrichment Score
	# sourya - commented since old bioconductor version does not support this function
	if (0) {	
		tsse <- TSSEscore(gal1, txs)
		capture.output(summary(tsse$TSS.enrichment.score), file=paste0(outdir, '/Summary_TSS_Enrichment_Score.txt'), append=FALSE)
	}


	# Heatmap and coverage curve for nucleosome positions
	# sourya - commented since old bioconductor version does not support this function
	if (0) {	
		nucleosome_positions_bamfiles <- file.path(outdir, c("NucleosomeFree.bam", "mononucleosome.bam", "dinucleosome.bam", "trinucleosome.bam"))
		plotfile <- paste0(outdir, '/Nucleosome_cumulative_Percentage.pdf')
		pdf(plotfile, width=6, height=4)
		if ((opt$RefGenome == 'hg19') | (opt$RefGenome == 'hg38')) {
			cumulativePercentage(nucleosome_positions_bamfiles[1:2], as(seqinfo(Hsapiens), "GRanges"))
		} else if (opt$RefGenome == 'mm10') {
			cumulativePercentage(nucleosome_positions_bamfiles[1:2], as(seqinfo(Mmusculus), "GRanges"))
		} else if (opt$RefGenome == 'mm9') {
			cumulativePercentage(nucleosome_positions_bamfiles[1:2], as(seqinfo(Mmusculus), "GRanges"))
		}
		dev.off()
		cat(sprintf("\n *** Performed function - Nucleosome_cumulative_Percentage ****  \n"))
	}

	# TSS statistics
	# sourya - commented since old bioconductor version does not support this function
	if (0) {
		TSS <- promoters(txs, upstream=0, downstream=1)
		TSS <- unique(TSS)
		# estimate the library size for normalization
		(librarySize <- estLibSize(nucleosome_positions_bamfiles))
		# calculate the signals around TSSs
		NTILE <- 101
		dws <- ups <- 1010
		sigs <- enrichedFragments(gal=objs[c("NucleosomeFree", "mononucleosome", "dinucleosome", "trinucleosome")], TSS=TSS, librarySize=librarySize, TSS.filter=0.5, n.tile = NTILE, upstream = ups, downstream = dws)
		# log2 transformed signals
		sigs.log2 <- lapply(sigs, function(.ele) log2(.ele+1))
		# plot heatmap
		plotfile <- paste0(outdir, '/Heatmap_signal_around_TSS.pdf')
		pdf(plotfile, width=6, height=4)
		featureAlignedHeatmap(sigs.log2, reCenterPeaks(TSS, width=ups+dws), zeroAt=.5, n.tile=NTILE)
		dev.off()
		cat(sprintf("\n *** Performed featureAlignedHeatmap **** \n"))

		# get signals normalized for nucleosome-free and nucleosome-bound regions
		out_Align_Distr <- featureAlignedDistribution(sigs, reCenterPeaks(TSS, width=ups+dws), zeroAt=.5, n.tile=NTILE, type="l",  ylab="Averaged coverage")
		# rescale the nucleosome-free and nucleosome signals to 0~1
		range01 <- function(x){(x-min(x))/(max(x)-min(x))}
		out_Align_Distr <- apply(out_Align_Distr, 2, range01)
		plotfile <- paste0(outdir, '/nucleosome_free_and_nucleosome_signals.pdf')
		pdf(plotfile, width=6, height=4)
		matplot(out_Align_Distr, type="l", xaxt="n", xlab="Position (bp)", ylab="Fraction of signal")
		axis(1, at=seq(0, 100, by=10)+1, labels=c("-1K", seq(-800, 800, by=200), "1K"), las=2)
		abline(v=seq(0, 100, by=10)+1, lty=2, col="gray")
		dev.off()
		cat(sprintf("\n *** Performed featureAlignedDistribution **** \n"))
	}

}	# paired end condition

# return to the original directory
setwd(currdir)

